components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +70,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
-        MDCButton *lastObject = [mutableButtons lastObject];
-        [lastObject removeFromSuperview];
-        [mutableButtons removeLastObject];
-      }
+- (void)removeObservers {
+  [self.imageView removeObserver:self forKeyPath:kMDCBannerViewImageViewImageKeyPath];
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary *)change
+                       context:(void *)context {
+  if (!context) {
+    if ([keyPath isEqualToString:kMDCBannerViewImageViewImageKeyPath]) {
+      self.imageView.hidden = (change[NSKeyValueChangeNewKey] == [NSNull null]);
     }
-    self.buttons = [mutableButtons copy];
   }
 }
 
-- (NSUInteger)numberOfButtons {
-  return self.buttons.count;
-}
+// TODO: KVO trailingButton.hidden and call `setNeedsUpdateConstraints:`.
 
 #pragma mark - UIView overrides
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size
-                                                     textLabel:self.textLabel
-                                                imageContainer:self.iconImageViewContainer
-                                                       buttons:self.buttons];
-  return [self.layout frameSize];
+  MDCBannerViewLayoutStyle style = [self styleForSizeToFit:size];
+  CGFloat frameHeight = 0.0f;
+  switch (style) {
+    case MDCBannerViewLayoutSingleLineStyle: {
+      frameHeight += kTopPaddingSmall + kBottomPadding;
+      CGSize textLabelSize = [self.textLabel sizeThatFits:CGSizeZero];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGFloat maximumHeight = MAX(textLabelSize.height, leadingButtonSize.height);
+      if (self.hasImage) {
+        maximumHeight = MAX(kImageViewSideLength, maximumHeight);
+      }
+      frameHeight += maximumHeight;
+      break;
+    }
+    case MDCBannerViewLayoutMultiLineAlignedButtonStyle: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGSize trailingButtonSize = [self.trailingButton sizeThatFits:CGSizeZero];
+      frameHeight += MAX(leadingButtonSize.height, trailingButtonSize.height);
+      break;
+    }
+    case MDCBannerViewLayoutMultiLineStackedButtonStyle: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGSize trailingButtonSize = [self.trailingButton sizeThatFits:CGSizeZero];
+      frameHeight +=
+          leadingButtonSize.height + trailingButtonSize.height + kButtonVerticalIntervalSpace;
+      break;
+    }
+  }
+  return CGSizeMake(size.width, frameHeight);
```

Comment:
I'm surprised that we still need custom logic in `sizeThatFits:` if we're supporting auto layout, are we able to convert this remaining logic to constraints?

Reply:
When `custom logic` is mentioned, do you mean the logic to calculate `style` or the entire logic here? For the entire logic, it is used to decided the view size based on given properties (label, image..in this case). For the `style` logic, it might not be easy and straightforward to do everything directly with constraints. For example, the spec requires the top padding to be 10 when all elements are able to fit on the same line while it to be 24 when element are not able to fit. So it is easier to update constraint based on this `style`. There might be a way to fully use constraints, but using some manual calculation here to get `style` would make it easier to set constraints.

Also because you mentioned `if we're supporting auto layout`, there is one thing worth noting here. `instrinsicContentSize` is not able to be implemented here because we don't have a `preferredContentWidth` property. So developer who uses auto layout needs to add specific constraints for its size.

Reply:
I think I understand. Though I'm still curious whether it's possible to make this "last mile" of layout considerations calculated via constraints as well, this is a small enough implementation detail that I won't block review of this PR on it.